templates/commands/render/templatesource/git.go

```
@@ -0,0 +1,218 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templatesource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/abcxyz/abc/templates/common"
+	"github.com/abcxyz/abc/templates/common/git"
+	"github.com/abcxyz/pkg/logging"
+	"golang.org/x/exp/slices"
+)
+
+// gitSourceParser implements sourceParser for downloading templates from a
+// remote git repo.
+type gitSourceParser struct {
+	// re is a regular expression that must have a capturing group for each
+	// group that is used in any of the "expansions" below. For example, if
+	// sshRemoteExpansion mentions "${host}", then re must have a group like
+	// "(?P<host>[a-zA-Z1-9]+)". See source.go for examples.
+	re *regexp.Regexp
+
+	// These fooExpansion strings are passed to
+	// https://pkg.go.dev/regexp#Regexp.Expand, which uses the syntax "$1" or
+	// "${groupname}" to refer to the values captured by the groups of the regex
+	// above.
+
+	// Example: `https://${host}/${org}/${repo}.git`
+	httpsRemoteExpansion string
+	// Example: `git@${host}:${org}/${repo}.git`
+	sshRemoteExpansion string
+	// Example: `${subdir}`
+	subdirExpansion string
+	// Example: `${version}`
+	versionExpansion string
+}
+
+func (g *gitSourceParser) sourceParse(ctx context.Context, src, protocol string) (templateDownloader, bool, error) {
+	match := g.re.FindStringSubmatchIndex(src)
+	if match == nil {
+		return nil, false, nil
```

Do we want to return an error here?